File: Reqnroll/Chapter15/15.7-TheStubDependencyPattern/After/WIMP.Specs/Support/PlaceOrderRequestObjectMother.cs
using WIMP.App.Models;
using WIMP.App.RestApi;

namespace WIMP.Specs.Support;

public class PlaceOrderRequestObjectMother
{
    private readonly List<PizzaItem> items = new();
    private readonly string email = DomainDefaults.CustomerEmail;
    private readonly string deliveryAddress = DomainDefaults.CustomerAddress;
    private DateTimeOffset? expectedDeliveryTime = null;

    public PlaceOrderRequestObjectMother()
    {
        // add default item
        items.Add(DomainDefaults.PizzaItemDefaultInstance());
    }

    public PlaceOrderRequest Build()
    {
        return new(items.ToArray(), email, deliveryAddress, ExpectedDeliveryTime: expectedDeliveryTime);
    }

    public PlaceOrderRequestObjectMother WithAdditionalItem(
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        var pizzaItem = DomainDefaults.PizzaItemDefaultInstance(name, size, style);
        items.Add(pizzaItem);
        return this;
    }

    public PlaceOrderRequestObjectMother WithItem(
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        items.Clear(); // remove existing items
        return WithAdditionalItem(name, size, style);
    }

    public PlaceOrderRequestObjectMother WithItems(int quantity,
        string? name = null, PizzaSize? size = null, PizzaStyle? style = null)
    {
        items.Clear(); // remove existing items
        for (int i = 0; i < quantity; i++)
        {
            WithAdditionalItem(name, size, style);
        }
        return this;
    }
    public PlaceOrderRequestObjectMother WithExpectedDeliveryTime(
        DateTimeOffset time)
    {
        expectedDeliveryTime = time;
        return this;
    }
}

File: CucumberJVM/Chapter16/16.3-EnsurePattern/After/src/test/java/com/wimp/app/specs/support/PlaceOrderRequestObjectMother.java
package com.wimp.app.specs.support;

import com.wimp.app.models.PizzaItem;
import com.wimp.app.models.PizzaSize;
import com.wimp.app.models.PizzaStyle;
import com.wimp.app.restapi.PlaceOrderRequest;

import java.util.ArrayList;
import java.util.List;

public class PlaceOrderRequestObjectMother {
    private final List<PizzaItem> items = new ArrayList<>();

    public PlaceOrderRequestObjectMother() {
        // add default item
        items.add(DomainDefaults.pizzaItemDefaultInstance());
    }

    public PlaceOrderRequest build() {
        return new PlaceOrderRequest(
            items.toArray(new PizzaItem[0]),
            DomainDefaults.CUSTOMER_ADDRESS,
            DomainDefaults.CUSTOMER_EMAIL);
    }

    public PlaceOrderRequestObjectMother withAdditionalItem(String name, PizzaSize size, PizzaStyle style) {
        PizzaItem pizzaItem = DomainDefaults.pizzaItemDefaultInstance(name, size, style);
        items.add(pizzaItem);
        return this;
    }

    public PlaceOrderRequestObjectMother withAdditionalItem() {
        return withAdditionalItem(null, null, null);
    }

    public PlaceOrderRequestObjectMother withItem(String name, PizzaSize size, PizzaStyle style) {
        items.clear(); // remove existing items
        return withAdditionalItem(name, size, style);
    }

    public PlaceOrderRequestObjectMother withItem(String name) {
        return withItem(name, null, null);
    }

    public PlaceOrderRequestObjectMother withItem() {
        return withItem(null, null, null);
    }

    public PlaceOrderRequestObjectMother withItems(int quantity, String name, PizzaSize size, PizzaStyle style) {
        items.clear(); // remove existing items
        for (int i = 0; i < quantity; i++) {
            withAdditionalItem(name, size, style);
        }
        return this;
    }

    public PlaceOrderRequestObjectMother withItems(int quantity) {
        return withItems(quantity, null, null, null);
    }

    public PlaceOrderRequestObjectMother withItems(int quantity, PizzaSize size) {
        return withItems(quantity, null, size, null);
    }
}
